ethereum/permissioning/src/main/java/tech/pegasys/pantheon/ethereum/permissioning/node/NodePermissioningController.java

```diff
@@ -25,22 +25,28 @@
   private static final Logger LOG = LogManager.getLogger();
 
   private final Optional<SyncStatusNodePermissioningProvider> syncStatusNodePermissioningProvider;
+  NodePermissioningProvider[] providers;
 
   public NodePermissioningController(
-      final SyncStatusNodePermissioningProvider syncStatusNodePermissioningProvider) {
+      final SyncStatusNodePermissioningProvider syncStatusNodePermissioningProvider,
+      final NodePermissioningProvider[] providers) {
+    this.providers = providers;
     this.syncStatusNodePermissioningProvider = Optional.of(syncStatusNodePermissioningProvider);
   }
 
-  public NodePermissioningController() {
-    this.syncStatusNodePermissioningProvider = Optional.empty();
-  }
-
   public boolean isPermitted(final EnodeURL sourceEnode, final EnodeURL destinationEnode) {
     LOG.trace("Checking node permission: {} -> {}", sourceEnode, destinationEnode);
 
-    return syncStatusNodePermissioningProvider
-        .map((provider) -> provider.isPermitted(sourceEnode, destinationEnode))
-        .orElse(true);
+    if (syncStatusNodePermissioningProvider.isPresent()
+        && !syncStatusNodePermissioningProvider.get().isPermitted(sourceEnode, destinationEnode)) {
+      return false;
+    } else {
+      for (int i = 0; i < providers.length; i++)
```

Comment:
If the array has the wrong order of providers (smart contract first, and then local config), we might end up with the wrong logic. Either we force the user to specify the providers by type and ensure we set them in the proper order, or we need to check that they are in the proper order.

Also, we need to think about the case when only one of the two types of permissioning is being used. E.g. only local config vs. only smart contract permissioning.

Reply:
Order doesn't matter from current thinking. In discussions (if memory serves) we decided we wanted to make sure we pass all permissioning filters to allow a connection rather than have one override the other, allowing the intersection of the 2 permissioning lists. The use case this currently misses is if a node is banned on the smart contract but allowed on the local contract which we discussed not doing for the moment as it allows the addition of nodes the network agreed shouldn't be there and we would wait for a real user to ask for it then consider it again.